doc/pbh/pbh-design.md

```diff
@@ -0,0 +1,770 @@
+        }
+    },
+    "PBH_TABLE": {
+        "pbh_table": {
+            "port_list": [
+                "Ethernet0",
+                "Ethernet4"
+            ],
+            "lag_list": [
+                "PortChannel0001",
+                "PortChannel0002"
+            ],
+            "description": "NVGRE and VxLAN"
+        }
+    }
+}
+```
+
+## 2.5 Flows
+
+### 2.5.1 PBH add
+
+![PBH add flow](images/pbh_add_flow.svg "Figure 3: PBH add flow")
+
+###### Figure 3: PBH add flow
+
+### 2.5.2 PBH remove
+
+![PBH remove flow](images/pbh_remove_flow.svg "Figure 4: PBH remove flow")
+
+###### Figure 4: PBH remove flow
+
+## 2.6 CLI
+
+### 2.6.1 Command structure
+
+**User interface**:
+```
+config
+|--- pbh
+     |--- table
+     |    |--- add <table_name> OPTIONS
+     |    |--- update <table_name> OPTIONS
+     |    |--- remove <table_name>
+     |
+     |--- rule
+     |    |--- add <rule_name> <table_name> OPTIONS
+     |    |--- update <rule_name> <table_name> OPTIONS
+     |    |--- remove <rule_name> <table_name>
+     |
+     |--- hash
+          |--- add <hash_name> OPTIONS
+          |--- update <hash_name> OPTIONS
+          |--- remove <hash_name>
+
+show
+|--- pbh
+     |--- table
+     |--- rule
+     |--- hash
+     |--- statistics
+```
+
+**Options:**
+
+_config pbh table add_
+1. -p|--port_list - port list
+2. -l|--lag_list - portchannel list
+3. -d|--description - table description
+
+_config pbh rule add_
+1. -p|--priority - rule priority
+2. -m|--match - match field
+3. -h|--hash_list - hash field list
+4. -a|--action=<set_ecmp_hash|set_lag_hash> - packet action
+5. -c|--counter=<enabled|disabled> - packet/byte counter
+
+_config pbh hash add_
+1. -f|--field - hash field
+3. -m|--mask - ip mask
+2. -s|--sequence - sequence id
+
+### 2.6.2 Usage examples
+
+#### 2.6.2.1 Config command group
+
+**The following command adds/updates/removes table:**
+```bash
+config pbh table add 'pbh_table' --port_list 'Ethernet0,Ethernet4' --lag_list 'PortChannel0001,PortChannel0002' \
+--description 'NVGRE and VxLAN'
+config pbh table update 'pbh_table' --port_list 'Ethernet0'
+config pbh table remove 'pbh_table'
+```
+
+**The following command adds/updates/removes rule:**
+```bash
+config pbh rule add 'vxlan' 'pbh_table' --priority 1 \
+--match gre_key 0x2500/0xffffff00 --match inner_ether_type 0x86dd/0xffff \
+--hash_list 'inner_ip_proto,inner_l4_dst_port,inner_l4_src_port,inner_dst_ipv6,inner_src_ipv6' \
+--action set_ecmp_hash --counter enabled
+config pbh rule update 'vxlan' 'pbh_table' --counter disabled
+config pbh rule remove 'vxlan' 'pbh_table'
+```
+
+**The following command adds/updates/removes hash:**
+```bash
+config pbh hash add 'inner_dst_ipv6' --field 'INNER_DST_IPV6' --mask 'FFFF::' --sequence 4
+config pbh hash update 'inner_dst_ipv6' --mask 'FFFF:FFFF::'
+config pbh hash remove 'inner_dst_ipv6'
+```
+
+#### 2.6.2.2 Show command group
+
+**The following command shows table configuration:**
```

Comment:
Is the desgin handles config validations? 
How about the config didn;t go through the ASIC? does it handle failures?

Reply:
@madhupalu In SONiC there is no response concept for CLI mechanism. Which means that we are limited in validation and sometimes the only way to get to know if the configuration was successful is to check the logs.

Reply:
You are right for CLI validations, however, there is a SAI Error handling HLD - https://github.com/Azure/SONiC/blob/master/doc/error-handling/error_handling_design_spec.md
you can refer this for SAI error handling for PBH SAI object failiures. It would be advice to list down what are the applications should listen/react to PBH failures?